File: java/csc460/searchproblems/MiceAndMeows.java
package csc460.searchproblems;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.TreeSet;
import java.util.Scanner;

import csc460.Board;
import csc460.BoardCoordinate;
import csc460.SearchState;
import java.awt.Color;

/**
 * Solves a Mice and Meows problem, where the agent (a cat) starts at a given spot then has to
 * eat up all the mice on a grid board and finally make it to the door. Water spots (puddles)
 * cost more for the cat to go into due to the cat not liking water.
 * See the loadBoardFile method for the board representation.
 * 
 * @author Hank Feild (hfeild@endicott.edu)
 */
public class MiceAndMeows implements SearchProblem {
    private SearchState startState;
    private BoardCoordinate exit;
    private BoardCoordinate start;
    // Used for determining next moves.
    private ArrayList<ArrayList<Character>> internalBoard;
    private HashMap<Character,Color> colorMap;
    private String heuristic;
    private TreeSet<BoardCoordinate> water;

    /**
     * Represents a maze state, where we only need to know the location of
     * the agent as well as details about that location (what it costs,
     * estimated distance to the exit, how the agent moved into the spot).
     */
    public class MiceAndMeowsState extends SearchState {
        private BoardCoordinate agentCoord;
        private TreeSet<BoardCoordinate> mice;
        private String action;
        private double cost, distance;

        /**
         * Initializes the maze state.
         * 
         * @param agentCoord Agent's location on the board. 
         * @param action The action that led to this state.
         * @param cost The cost of moving to this state.
         * @param distance The estimated distance to the exit.
         */
        public MiceAndMeowsState(BoardCoordinate agentCoord, TreeSet<BoardCoordinate> mice,
                String action, double cost,
                double distance){
            this.agentCoord = agentCoord;
            this.action = action;
            this.cost = cost;
            this.distance = distance;
            this.mice = mice;
        }

        /**
         * @return The location of the agent.
         */
        @Override
        public BoardCoordinate getAgentCoordinates() {
            return agentCoord;
        }

        /**
         * @return The mice locations in a set.
         */
        public TreeSet<BoardCoordinate> getMiceCoordinates() {
            return mice;
        }

        /**
         * @return The cost to reach this state.
         */
        @Override
        public double getCost() {
            return cost;
        }

        /**
         * @return The estimated distance to the goal.
         */
        @Override
        public double getDistance() {
            return distance;
        }

        /**
         * @return The action that led to this state.
         */
        @Override
        public String getAction() {
            return action;
        }

        /**
         * @return A string representation of the state.
         */
        public String toString(){
            StringBuffer description = new StringBuffer(agentCoord.toString());
            for(BoardCoordinate coord : mice){
                description.append(coord.toString());
            }
            return description.toString();
        }

    }

    /**
     * Creates a map of character codes (from the board file) to colors.
     */
    public MiceAndMeows() {
        colorMap = new HashMap<Character,Color>();
        colorMap.put(' ', Color.WHITE);
        colorMap.put('s', Color.GREEN);
        colorMap.put('e', Color.RED);
        colorMap.put('w', Color.BLACK);
    }

    /**
     * Sets the heuristic to use to estimate the distance to a goal state. 
     * The following are currently supported:
     * 
     *   - manhattan
     *   - euclidean
     *   - mice-remaining
     */
    @Override
    public void setHeuristic(String heuristic) {
        this.heuristic = heuristic;
    }

    /**
     * @return The starting state.
     */
    @Override
    public SearchState getStartState() {
        return startState;
    }

    /**
     * Determines if the given state is a goal state.
     * 
     * @param state The state to test.
     * @return True if the agent has reached the exit.
     */
    @Override
    public boolean isGoal(SearchState state) {
        return ((MiceAndMeowsState) state).getMiceCoordinates().isEmpty() && state.getAgentCoordinates().equals(exit);
    }

    /**
     * Estimates the distance from the given coordinate to a goal state using the
     * heuristic set for the instance.
     * 
     * @param coord A spot on the board.
     * 
     * @return An estimate of the distance from the given coordinate to the 
     *         exit.
     */
    public double getDistance(BoardCoordinate coord, TreeSet<BoardCoordinate> mice){
        if(heuristic.equals("manhattan")){
            return Math.abs(coord.x - exit.x) + Math.abs(coord.y - exit.y);
        } else if(heuristic.equals("euclidean")){
            return Math.sqrt(Math.pow(coord.x-exit.x, 2) + 
                             Math.pow(coord.y-exit.y, 2));

        } else if(heuristic.equals("mice-remaining")){
            if(mice.size() > 0){
                return mice.size();
            } else {
                return Math.abs(coord.x - exit.x) + Math.abs(coord.y - exit.y);
            }
        } else {
            return 0;
        }
    }

    /**
     * Finds all of the states that can be reached from the given state.
     * 
     * @param state The state to find successors of.
     * @return A collection of the states to the left, up, right, and down of 
     *         the given state.
     */
    @Override
    public Iterable<SearchState> getSuccessors(SearchState state) {
        ArrayList<SearchState> successors = new ArrayList<SearchState>();

        BoardCoordinate curCoords = state.getAgentCoordinates();

        // This speeds up checking the four directions (so we can use a loop).
        String[] actions = {"left", "up", "right", "down"};
        BoardCoordinate[] successorCoords = {
            new BoardCoordinate(curCoords.x-1, curCoords.y), // left
            new BoardCoordinate(curCoords.x, curCoords.y-1), // up
            new BoardCoordinate(curCoords.x+1, curCoords.y), // right
            new BoardCoordinate(curCoords.x, curCoords.y+1) // down
        };

        // Look in each direction around the agent. If it's in bounds, add it to the list
        // of successors.
        for(int i = 0; i < actions.length; i++){
            if(successorCoords[i].y >= 0 && 
                    successorCoords[i].y < internalBoard.size() &&
                successorCoords[i].x >= 0 && 
                    successorCoords[i].x < internalBoard.get(0).size()){

                // Set the cost: 6 for water, 1 for every other spot.
                int cost = 1;
                if (internalBoard.get(successorCoords[i].y).get(successorCoords[i].x) == 'w' ){
                    cost = 6;
                }

                // Update the list of mice if one is colocated with this spot.
                TreeSet<BoardCoordinate> mice = new TreeSet<BoardCoordinate>(((MiceAndMeowsState) state).getMiceCoordinates());
                if(mice.contains(successorCoords[i])){
                    mice.remove(successorCoords[i]);
                }

                successors.add(new MiceAndMeowsState(successorCoords[i], mice, actions[i], 
                    cost, getDistance(successorCoords[i], mice)));
            }
        }

        return successors;
    }

    /**
     * Parses a Mice and Meows file. Should contain one character per spot.
     * Here are the character codes:
     *   s -- the agent's starting position
     *   e -- exit (if reached and all mice have been collected, the goal state has been reached)
     *   w -- water (cost: 6)
     *   m -- mouse (all of these must be collected, plus the exit reached in order for the goal test to pass)
     *   (blank) -- a spot the agent may enter (cost: 1)
     * 
     * @param filename The name of the maze file to load.
     * @return A color mapped version of the maze.
     */
    @Override
    public Board loadBoardFile(String filename) throws FileNotFoundException {
        Scanner reader = new Scanner(new File(filename));
        String line;
        char spot;
        int x = 0, y = 0;
        Board board;
        TreeSet<BoardCoordinate> mice = new TreeSet<BoardCoordinate>();
        internalBoard = new ArrayList<ArrayList<Character>>();

        while(reader.hasNextLine()){
            line = reader.nextLine();
            ArrayList<Character> row = new ArrayList<Character>();

            for(x = 0; x < line.length(); x++){
                spot = line.charAt(x);

                // Add the spot to the internal representation.
                row.add(spot);
                if(spot == 's'){
                    start = new BoardCoordinate(x, y);
                } else if(spot == 'e') {
                    exit = new BoardCoordinate(x, y);
                } else if(spot == 'm') {
                    mice.add(new BoardCoordinate(x,y));
                }
            }
            internalBoard.add(row);
            y++;
        }
        reader.close();

        // Initializes the start state.
        startState = new MiceAndMeowsState(start, mice, "", 0.0, getDistance(start, mice));

        board = new Board(internalBoard.size(), internalBoard.get(0).size());
        // Map the spot to a color and add that to the external
        // board representation.

        y = 0;
        for(ArrayList<Character> row : internalBoard){
            x = 0;
            for(Character c : row){
                board.setColor(new BoardCoordinate(x,y), colorMap.get(c.charValue()));
                if(c == ' '){
                    System.out.print('_');
                } else {
                    System.out.print(c);
                }
                x++;
            }
            System.out.println();
            y++;
        }

        return board;
    }
    
}

File: python/mice_and_meows.py
# EXCLUDE FILE
'''
File:    mice_and_meows.py
Author:  Prof Feild
Purpose: Defines the Maze problem where an agent must navigate from a start
         position to an exit position in a grid-based maze.
'''

from math import sqrt
import sys
import problem

class MiceAndMeowState:
    '''Represents a state in the Mouse and Meows problem, which is the
    agent's position and the position of every mouse that has yet to be
    caught.
    '''
    def __init__(self, agentLocation, miceLocations):
        '''
        Parameters:
            agentLocation ((int,int)): The location of the agent in the maze.
            miceLocations (((int, int), ...)): A tuple of the locations of the mice in the maze.
        '''
        self.agentLocation = agentLocation
        self.miceLocations = miceLocations

    def __eq__(self, other):
        '''
        Returns (bool): Whether two states are equal; they are equal if the
            agent and mice are in the same locations in both states.
        '''
        return (self.agentLocation == other.agentLocation and 
            self.miceLocations == other.miceLocations)
    
    def __hash__(self):
        '''
        Returns (int): The hash of a state, which is the hash of the agent's
            location and the hash of the set of mice locations.
        '''
        return hash((self.agentLocation, self.miceLocations))
    

class MiceAndMeows(problem.Problem):
    '''
    The search problem: Of Mice and Meows. The agent is a cat and the goal is to
    catch all of the mice and make it to the exit. The cost of moving is 1,
    except for moving into water, which costs 6.
    '''
    heuristics = set(('manhattan', 'euclidean', 'mice-remaining','totalDistance','closestMice'))
    visited = []
    def __init__(self, boardFilename=None, heuristic=None):
        '''
        Parameters:
            boardFilename (str): The filename of the board to load.
            heuristic (str): The heuristic to use for informed search.
        '''  
        if boardFilename == None:
            self.board = []
            self.start = MiceAndMeowState((-1,-1), set())
            self.exit = (-1,-1)
        else:
            self.loadFile(boardFilename)

        heuristics = {
            'manhattan': self.manhattanDistance,
            'euclidean': self.euclideanDistance,
            'mice-remaining': self.miceRemainingDistance,
            'totalDistance' : self.heuristicTotalDistance,
            'closestMice' : self.heuristicClosestMice
        }
        if heuristic is None:
            self.heuristic = self.defaultHeuristic
        elif heuristic in heuristics:
            self.heuristic = heuristics[heuristic]
        else:
            sys.exit('Invalid heuristic: ' + heuristic)

        print('Heuristic:', heuristic)

    def loadFile(self, boardFilename):
        '''
        Parses a Mouse and Meows board file. Should contain one character per spot. The cat and mice
        positions are encoded in a state, while the exit and water positions are encoded in
        the problem instance itself (since those are fixed).

        Here are the character codes:
            s -- the agent's starting position
            e -- exit (if reached and all mice are gone, the goal state has been reached)
            w -- water (cost: 6)
            m -- mouse (0 or more)
            (space) -- an open spot (cost: 1)

        Parameters:
            boardFilename (str): The path to a file containing one character per spot on the board.
        '''
        self.board = []
        agentLocation = (-1,-1)
        miceLocations = []
        for i,row in enumerate(open(boardFilename)):
            self.board.append([])
            for j,col in enumerate(row.rstrip('\n')):
                self.board[-1].append(col)
                if col == 's':
                    agentLocation = (i,j)
                elif col == 'e':
                    self.exit = (i,j)
                elif col == 'm':
                    miceLocations.append((i,j))

        self.start = MiceAndMeowState(agentLocation, tuple(miceLocations))
        print('Starting state:')
        for row in self.board:
            print(''.join(['_' if c == ' ' else c for c in row]))

    def successors(self, state):
        '''
        Produces a list of spots -- ()  -- that the agent can move into, along
        with the spot's cost and estimate of how far it is from the exit.
        Returned as a list of tuples:

            [
                (move, MiceAndMeowsState, cost, dist), ...
            ]

        Parameters:
            state (MiceAndMeowsState): The current state of the problem.

        Returns (list((str, MiceAndMeowsState, float, float))): A list of the
            states that can be reached from the given state. Each item is a
            4-tuple: (action, new state, cost of action, estimated cost to
            goal).
        '''
        (i,j) = state.agentLocation
        successors = []
        potentialSuccessorSpots = (
            ('left', i,j-1),
            ('up', i-1,j), 
            ('right', i,j+1), 
            ('down', i+1,j) 
        )

        ## some kind of a list to keep track of visited nodes and t
    
        for move,i,j in potentialSuccessorSpots:
            if( i >= 0 and i < len(self.board) and 
                j >= 0 and j <len(self.board[i])):

                cost = 6 if self.board[i][j] == 'w' else 1

                mice = state.miceLocations
                # If the agent is moving onto a mouse, remove it from the set of mice.,
                if (i,j) in mice:
                    mice = tuple([m for m in mice if m != (i,j)])

                successorState = MiceAndMeowState((i,j), mice)

                if successorState not in self.visited:
                    successors.append((move, successorState, cost, self.getDistance(successorState)))
                    self.visited.append(successors)

        return successors

    def getDistance(self, state):
        '''
        Estimates the distance from the state to a goal state using whatever the current heuristic is.

        Parameters:
            state (MiceAndMeowsState): The state to check.
            
        Returns (float): The estimated distance from the given state to a goal state. This should
            use whatever heuristic is currently set for the problem.
        '''
        return self.heuristic(state)

    def isGoal(self, state):
        '''
        The goal is reached when the agent has caught all the mice and is at the exit.

        Parameters:
            state (MiceAndMeowsState): The state to check.

        Returns (bool): True if the given state is a goal state, False otherwise.
        '''     
        return state.agentLocation == self.exit and len(state.miceLocations) == 0

    def defaultHeuristic(self, state):
        '''Uses 0 for all estimates.

        Parameters:
            state (MiceAndMeowsState): The state to check.

        Returns (float): The default heuristic value for the given state: 0.
        '''
        return 0

    def manhattanDistance(self, state):
        '''
        Parameters:
            state (MiceAndMeowsState): The state to check.

        Returns (float): The Manhattan distance from the given state to the exit
            location.
        '''
        return abs(self.exit[0]-state.agentLocation[0]) + abs(self.exit[1]-state.agentLocation[1])

    def euclideanDistance(self, state):
        '''
        Parameters:
            state (MiceAndMeowsState): The state to check.

        Returns (float): The Euclidean distance from the given state to the exit
            location.
        '''
        return float(sqrt(pow(self.exit[0]-state.agentLocation[0], 2) + pow(self.exit[1]-state.agentLocation[1], 2)))
    
    ## 
    def heuristicClosestMice(self,state):
        if len(state.miceLocations) == 0:
            return self.manhattanDistance(state)
        lowestDistance = float('inf')
        closestMouse = None
        for x,y in state.miceLocations:
            currentDistance = (abs(x - state.agentLocation[0])) + (abs(y - state.agentLocation[1]))
            if currentDistance < lowestDistance:
                lowestDistance = currentDistance
                closestMouse = (x, y)
        # h = distance to nearest mouse + Manhattan distance from that mouse to exit
        # Still admissible: agent must visit at least the closest mouse, then reach the exit
        distToExit = abs(self.exit[0] - closestMouse[0]) + abs(self.exit[1] - closestMouse[1])
        return lowestDistance + distToExit

    ##### second heuristic
    def heuristicTwo(self,state):
        return        
    
    ## This would work: Distance of all the mices together if there are mices, if not we caluclate the manhatten distance to the goal
    def heuristicTotalDistance(self,state):
        if len(state.miceLocations) == 0:
            return self.manhattanDistance(state)
        return(self.distanceBetweenMice(state))
    
    def distanceBetweenMice(self,state):
        totalDistance = 0
        index = 0
        for x,y in state.miceLocations:
            if index >= len(state.miceLocations) - 1:
                return totalDistance
            else:
                totalDistance += abs(x - state.miceLocations[index + 1][0]) + abs(y - state.miceLocations[index + 1][1])
                index += 1
    
    
    def miceRemainingDistance(self, state):
        '''
        Parameters:
            state (MiceAndMeowsState): The state to check.

        Returns (float): The number of mice remaining; if 0, uses the Manhattan
            distance from the given state to the exit location.
        '''
        if len(state.miceLocations) > 0:
            return len(state.miceLocations)
        return self.manhattanDistance(state)
